core/src/HIP/Kokkos_HIP_Instance.hpp

```diff
@@ -65,6 +65,93 @@ HIP::size_type *hip_internal_scratch_flags(const HIP &instance,
 
 //----------------------------------------------------------------------------
 
+// Helper to protect a shared resource from being used by multiple streams
+// simultaneously.
+// If used properly, only one stream at a time will be able to use the shared
+// resource.
+// This helper should be used in a thread-safe way.
+//
+// Typical usage:
+// @code
+// auto lock_for_acquisition = shared_resource.acquire();
+//
+// ... do stuff on the shared resource ...
+//
+// shared_resource.release(std::move(lock_for_acquisition), stream);
+//
+// ...
+//
+// auto lock_for_involvement = shared_resource.lock();
+// hipStreamSynchronize(stream);
+// shared_resource.check_if_involved(std::move(lock_for_involvement), stream);
+// hipStreamDestroy(stream);
+// @endcode
+struct SharedResourceLocking {
+  bool m_need_sync = false;
+  std::mutex m_mutex{};
+  hipEvent_t m_event   = nullptr;
+  hipStream_t m_stream = nullptr;
+
+  // Acquire the right to interact in a thread-safe way.
+  [[nodiscard]] auto lock() { return std::unique_lock<std::mutex>{m_mutex}; }
+
+  // The event is created for the current device. The instance is locked first.
+  void initialize() {
+    auto lock = this->lock();
+    if (!m_event) {
+      KOKKOS_IMPL_HIP_SAFE_CALL(
+          hipEventCreateWithFlags(&m_event, hipEventDisableTiming));
+    }
+  }
+
+  // Destroying an event can be done even if it is not bound to the current
+  // device.
+  void finalize() {
+    auto lock = this->lock();
+    if (m_event) {
+      KOKKOS_IMPL_HIP_SAFE_CALL(hipEventDestroy(m_event));
+    }
+  }
+
+  SharedResourceLocking()                                   = default;
+  SharedResourceLocking(SharedResourceLocking const &other) = delete;
+  SharedResourceLocking &operator=(SharedResourceLocking const &other) = delete;
+  SharedResourceLocking &operator=(SharedResourceLocking &&other)      = delete;
+
```

Comment:
The rule of 5 suggests to also be explicit about the destructor.
```suggestion
  SharedResourceLocking()                                   = default;
  SharedResourceLocking(SharedResourceLocking const &other) = delete;
  SharedResourceLocking &operator=(SharedResourceLocking const &other) = delete;
  SharedResourceLocking &operator=(SharedResourceLocking &&other)      = delete;
 ~SharedResourceLocking() = default;
```